File: packages/cpp/examples/optimizer_artifact.cpp
#include "axllm/axllm.hpp"
#include <iostream>

struct FakeOptimizer : axllm::OptimizerEngine {
  std::string name() const override { return "fixture"; }
  std::string version() const override { return "1"; }
  axllm::Value optimize(axllm::Value) override {
    return axllm::object({
      {"componentMap", axllm::object({{"qa::instruction", "Prefer artifact-backed answers."}})},
      {"metadata", axllm::object({
        {"evidence", axllm::object({{"avg", 1}})},
        {"provenance", axllm::object({{"sourceProgramKind", "axgen"}})}
      })}
    });
  }
};

static bool has_instruction(const axllm::AxGen& gen, const std::string& value) {
  axllm::Value components = gen.get_optimizable_components();
  for (int i = 0; ; ++i) {
    axllm::Value item = axllm::Core::get(components, i);
    if (axllm::Core::truthy(axllm::Core::is_none(item))) break;
    if (axllm::equal(axllm::Core::get(item, "id"), "qa::instruction") &&
        axllm::equal(axllm::Core::get(item, "current"), value)) return true;
  }
  return false;
}

int main() {
  axllm::AxGen qa = axllm::ax("question:string -> answer:string", axllm::object({{"id", "qa"}, {"instruction", "Base."}}));
  FakeOptimizer engine;
  axllm::Value artifact = qa.optimize_with(engine, axllm::Value::array(), axllm::object({{"apply", false}}));
  if (!has_instruction(qa, "Base.")) return 1;
  qa.apply_optimization(axllm::Value(axllm::stringify(artifact)));
  if (!has_instruction(qa, "Prefer artifact-backed answers.")) return 2;
  std::cout << "cpp-optimizer-artifact-ok\n";
}

File: packages/python/examples/optimizer_artifact.py
import json

from axllm import OptimizerEngine, ax


class FakeOptimizer(OptimizerEngine):
    name = "fixture"
    version = "1"

    def optimize(self, request, evaluator=None):
        return {
            "componentMap": {"qa::instruction": "Prefer artifact-backed answers."},
            "metadata": {
                "evidence": {"avg": 1},
                "provenance": {"sourceProgramKind": "axgen"},
            },
        }


qa = ax("question:string -> answer:string", {"id": "qa", "instruction": "Base."})
artifact = qa.optimize_with(FakeOptimizer(), [], {"apply": False})
assert any(item["id"] == "qa::instruction" and item["current"] == "Base." for item in qa.get_optimizable_components())
qa.apply_optimization(json.dumps(artifact))
assert any(
    item["id"] == "qa::instruction" and item["current"] == "Prefer artifact-backed answers."
    for item in qa.get_optimizable_components()
)
print("python-optimizer-artifact-ok")
